src/fastfilter.h

```diff
@@ -43,47 +45,74 @@ class CFastFilter
 
     CFastFilter()
     {
+        static_assert((HASHES > 1) && (HASHES <= 16), "HASHES must be between 2 and 16 inclusive");
         FastRandomContext insecure_rand;
         vData.resize(FILTER_BYTES);
-        // by sampling from a different part of the uint256, we make it harder for an attacker to generate collisions
-        grabFrom = (1 + (insecure_rand.rand32() % 6)) * 4; // should be 4 byte aligned for speed
+        GetRandBytes(salt.begin(), sizeof(uint256));
     }
 
     // returns true IF this function made a change (i.e. the value was previously not set).
     bool checkAndSet(const uint256 &hash)
     {
-        const unsigned char *mem = hash.begin();
-        const uint32_t *pos = (const uint32_t *)&(mem[grabFrom]);
-        uint32_t idx = ((*pos) ^ mem[0]) & (FILTER_SIZE - 1);
-
-        bool ret = vData[idx >> 3] & (1 << (idx & 7));
-        vData[idx >> 3] |= (1 << (idx & 7));
-        if (ret)
-            conflicts++;
-        return !ret;
+        const uint32_t *saltPos = (const uint32_t *)salt.begin();
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        bool unset = 0; // If any position is not set, then this will be true
+        for (unsigned int i = 0; i < HASHES / 2; i++, pos++, saltPos++)
+        {
+            uint32_t val = *pos ^ *saltPos;
+            uint32_t idx = val & (FILTER_SIZE - 1);
+            val = __builtin_bswap32(val);
+            uint32_t idx2 = val & (FILTER_SIZE - 1);
+
+            unset |= (0 == (vData[idx >> 3] & (1 << (idx & 7))));
+            unset |= (0 == (vData[idx2 >> 3] & (1 << (idx2 & 7))));
+
+            vData[idx >> 3] |= (1 << (idx & 7));
+            vData[idx2 >> 3] |= (1 << (idx2 & 7));
+        }
+        return unset;
     }
 
     void insert(const uint256 &hash)
     {
-        const unsigned char *mem = hash.begin();
-        const uint32_t *pos = (const uint32_t *)&(mem[grabFrom]);
-        uint32_t idx = ((*pos) ^ mem[0]) & (FILTER_SIZE - 1);
-        vData[idx >> 3] |= (1 << (idx & 7));
+        const uint32_t *saltPos = (const uint32_t *)salt.begin();
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        for (unsigned int i = 0; i < HASHES / 2; i++, pos++, saltPos++)
+        {
+            uint32_t val = *pos ^ *saltPos;
+            uint32_t idx = val & (FILTER_SIZE - 1);
+            val = __builtin_bswap32(val);
```

Comment:
this is a single instruction on an intel machine.  It is OK to reuse bits -- after all the traditional bloom filter reuses all the input bits in every hash function.

Reply:
But the key here is `all` the bits. Do you think this will be faster compared to just proper slicing?

Reply:
All the bits are used if you choose all the hashes.  Which we do.  If you choose fewer hashes, you are making a tradeoff of performance verses deliberate collision attacks IF and only if this structure is used with data specified by an external source.  Even in that case, what are the repercussions of collisions?  Could be minimal.  I feel like you are rejecting for example the radix tree data structure because it's possible to insert data that turns that tree into a list.  But there are many applications of radix trees where we have strong knowledge that the data won't do that.

Reply:
unrelated point: we sure that `__builtin_bswap32` compiler functions  is always available?

Reply:
It's a gcc extension, so its not always available.  However, many compilers support gcc extensions, and its compiling on all of our architectures.  If we port to an unsupported compiler we will need to add a compatibility macro at that time.